<a href="https://colab.research.google.com/github/TsamayaDesigns/codeDivision-data-with-python/blob/main/Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning
---



In order to effectively analyse a dataset, often we need to prepare it first.
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g.
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted*

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_excel_data():
  url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
  df = pd.read_excel(url)
  return df

data = get_excel_data()

# Interrogate data
def investigate_data(data):
  # 1. Display first 5 rows of data
  orig_data = data.head()

  # 2. Sort the data by Happiness Score in ascending order & display the sorted table
  sorted_data = data.sort_values("Happiness Score", ascending = True)

  return {
      "1. Display first 5 rows of data": orig_data,
      "2. Display the sorted data - (by Happiness Score, sorted: ascending order)": sorted_data,
  }

statistics = investigate_data(data)

print("\nExercise 1 - Get data & Sort by Happiness Score\n")
for key, value in statistics.items():
  print(f"{key}:\n{value}\n")



Exercise 1 - Get data & Sort by Hapiness Score

1. Display first 5 rows of data:
       Country          Region  Happiness Rank  Happiness Score  Standard Error  Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  Trust (Government Corruption)  Generosity  Dystopia Residual
0  Switzerland  Western Europe               1            7.587         0.03411                   1.39651  1.34951                   0.94143  0.66557                        0.41978     0.29678            2.51738
1      Iceland  Western Europe               2            7.561         0.04884                   1.30232  1.40223                   0.94784  0.62877                        0.14145     0.43630            2.70201
2      Denmark  Western Europe               3            7.527         0.03328                   1.32548  1.36058                   0.87464  0.64938                        0.48357     0.34139            2.49204
3       Norway  Western Europe               4            7.522       

### Exercise 2 - sort by multiple columns, display the first 5 rows
---

1. Sort the data by Health (Life Expectancy) and Economy (GDP per Capita) in ascending order
2. display the first 5 rows of sorted data

**Test output**:  
Records 122, 127, 147, 100, 96

In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_excel_data():
  url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
  df = pd.read_excel(url)
  return df

data = get_excel_data()

# Interrogate data
def investigate_data(data):
  # 1. Sort the data by Health (Life Expectancy) and Economy (GDP per Capita) in ascending order
  sort_data = data.sort_values(by = ["Health (Life Expectancy)", "Economy (GDP per Capita)"], ascending = [True, True])

  # 2. Display the first 5 rows of sorted data
  sorted_data_hd = sort_data.head()

  return {
      "1. Display Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order (first 5 rows of sorted data) ": sorted_data_hd,
  }

statistics = investigate_data(data)

print("\nExercise 2 - Sort by multiple columns & display the first 5 rows\n")
for key, value in statistics.items():
  print(f"{key}:\n{value}\n")


Exercise 2 - Sort by multiple columns & display the first 5 rows

1. Display Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order (first 5 rows of sorted data) :
                      Country              Region  Happiness Rank  Happiness Score  Standard Error  Economy (GDP per Capita)   Family  Health (Life Expectancy)  Freedom  Trust (Government Corruption)  Generosity  Dystopia Residual
122              Sierra Leone  Sub-Saharan Africa             123            4.507         0.07068                   0.33024  0.95571                   0.00000  0.40840                        0.08786     0.21488            2.51009
127                  Botswana  Sub-Saharan Africa             128            4.332         0.04934                   0.99355  1.10464                   0.04776  0.49495                        0.12474     0.10461            1.46181
147  Central African Republic  Sub-Saharan Africa             148            3.678         0.06112         

### Exercise 3 - sorting in descending order
---

Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_excel_data():
  url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
  df = pd.read_excel(url)
  return df

data = get_excel_data()

# Interrogate data
def investigate_data(data):
  # 1. Sort the data by Freedom and Trust (Government Corruption) in descending order. Show the Country and Region only for the last five rows
  sort_data = data.sort_values(by = ["Freedom", "Trust (Government Corruption)"], ascending = False)[["Country", "Region"]].tail()

  return {
      "1. Display Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order": sort_data,
  }

statistics = investigate_data(data)

print("\nExercise 3 - Sorting in descending order\n")
for key, value in statistics.items():
  print(f"{key}:\n{value}\n")


Exercise 3 - Sorting in descending order

1. Display Sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order:
                    Country                           Region
136                  Angola               Sub-Saharan Africa
117                   Sudan               Sub-Saharan Africa
95   Bosnia and Herzegovina       Central and Eastern Europe
101                  Greece                   Western Europe
111                    Iraq  Middle East and Northern Africa



# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv
2. check if any NA values exist in the dataframe and print the result
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)



In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Interrogate data
def investigate_data(data):
  # 1. Check if any NA values exist in the dataframe and print the result
  check_data = data.isnull().values.any()

  # 2. Use df.info() to see which columns have null entries
  null_entries = data.info()

  return {
      "1. Check if any NA values exist in the dataframe and print the result": check_data,
      "2. Use df.info() to see which columns have null entries": null_entries,
  }

statistics = investigate_data(data)

print("\nExercise 4 - Check for null values\n")
for key, value in statistics.items():
  print(f"{key}:\n{value}\n")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB

Exercise 4 - Check for null values

1. Check if any NA values exist in the dataframe and print the result:
True

2. Use df.info() to see which columns have 

### Exercise 5 - remove null values
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Interrogate data
def investigate_data(data):
  # 1. Display whole dataframe, with all the NA values"
  full_df = data.loc[:]

  # 2. Remove rows with NA values for "life_satisfaction"
  new_life_satis = data.dropna(subset = ["life_satisfaction"])

  # 3. Remove all NA values across whole dataframe
  rem_all_na = data.dropna()

  return {
      "1. Display whole dataframe, with all the NA values": full_df,
      "2. Remove rows with NA values for \"life_satisfaction\"": new_life_satis,
      "3. Remove all NA values across whole dataframe": rem_all_na,
  }

statistics = investigate_data(data)

print("\nExercise 5 - Remove null values\n")
for key, value in statistics.items():
  print(f"{key}:\n{value}\n")



Exercise 5 - Remove null values

1. Display whole dataframe, with all the NA values:
           code                  area        date  median_salary  life_satisfaction mean_salary recycling_pct  population_size  number_of_jobs  area_size  no_of_houses  borough_flag
0     E09000001        city of london  1999-12-01        33020.0                NaN       48922             0           6581.0             NaN        NaN           NaN             1
1     E09000002  barking and dagenham  1999-12-01        21480.0                NaN       23620             3         162444.0             NaN        NaN           NaN             1
2     E09000003                barnet  1999-12-01        19568.0                NaN       23128             8         313469.0             NaN        NaN           NaN             1
3     E09000004                bexley  1999-12-01        18621.0                NaN       21386            18         217458.0             NaN        NaN           NaN             1
4   

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])`

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01*

In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Interrogate data
def investigate_data(data):
  # 1. Display whole dataframe, with all the NA values"
  full_df = data.loc[:]

  # 2. Remove duplicate "area" entries
  rem_dupl_area = data.drop_duplicates(subset = ["area"])

  # 3. Remove rows with NA values in the "median_salary" column
  rem_na_med_sal = rem_dupl_area.dropna(subset = ["median_salary"])

  return {
      "1. Display whole dataframe, with all the NA values": full_df,
      "2. Remove duplicate \"area\" entries": rem_dupl_area,
      "3. Remove rows with NA values in the \"median_salary\" column": rem_na_med_sal,
  }

statistics = investigate_data(data)

print("\nExercise 6 - Removing duplicate entries\n")
for key, value in statistics.items():
  print(f"{key}:\n{value}\n")



Exercise 6 - Removing duplicate entries

1. Display whole dataframe, with all the NA values:
           code                  area        date  median_salary  life_satisfaction mean_salary recycling_pct  population_size  number_of_jobs  area_size  no_of_houses  borough_flag
0     E09000001        city of london  1999-12-01        33020.0                NaN       48922             0           6581.0             NaN        NaN           NaN             1
1     E09000002  barking and dagenham  1999-12-01        21480.0                NaN       23620             3         162444.0             NaN        NaN           NaN             1
2     E09000003                barnet  1999-12-01        19568.0                NaN       23128             8         313469.0             NaN        NaN           NaN             1
3     E09000004                bexley  1999-12-01        18621.0                NaN       21386            18         217458.0             NaN        NaN           NaN           

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Interrogate data
def normalise_income(data):
  if data["median_salary"] < 27441:
    return 1
  elif data["median_salary"] < 30932:
    return 2
  else:
    return 3

data["median_salary"] = data.apply(normalise_income, axis = 1)

print("\nExercise 7 - Normalise data set\n")
print(data.loc[:])




Exercise 7 - Normalise data set

           code                  area        date  median_salary  life_satisfaction mean_salary recycling_pct  population_size  number_of_jobs  area_size  no_of_houses  borough_flag
0     E09000001        city of london  1999-12-01              3                NaN       48922             0           6581.0             NaN        NaN           NaN             1
1     E09000002  barking and dagenham  1999-12-01              1                NaN       23620             3         162444.0             NaN        NaN           NaN             1
2     E09000003                barnet  1999-12-01              1                NaN       23128             8         313469.0             NaN        NaN           NaN             1
3     E09000004                bexley  1999-12-01              1                NaN       21386            18         217458.0             NaN        NaN           NaN             1
4     E09000005                 brent  1999-12-01       

### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [ ]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Interrogate data
# 1. Use df.describe() to find the median, upper quartile and maximum for the "number_of_jobs" column
stats = data["number_of_jobs"].describe()

# Median
median = stats["50%"]

# Upper Quartile
upper_quartile = stats["75%"]

#  Maximum
max = stats["max"]

print("\nExercise 8 - Normalise the number of jobs column")
print(f"\nUse .describe() to find the median, upper quartile and maximum for the \"number_of_jobs\" column:\n{stats}\n")
print(f"\nDisplay Median, Upper Quartile and Maximum:\nMedian: {median}\nUpper Quartile: {upper_quartile}\nMaximum: {max}\n")

# 2. Create a function called normalise_jobs(df) that will return 1 if the "number_of_jobs" is below the median, 2 if the "number_of_jobs" is below the upper quartile or 3 otherwise.
def normalise_jobs(data):
  if data["number_of_jobs"] < median:
    return 1
  elif data["number_of_jobs"] < upper_quartile:
    return 2
  else:
    return 3

# 3. Normalise the "number_of_jobs" column by applying the function normalise_jobs.

data["number_of_jobs"] = data.apply(normalise_jobs, axis = 1)

print(f"\nDisplay normalised \"number_of_jobs\" column:\n{data.loc[:]}\n")

stats2 = data["number_of_jobs"].describe()
print(f"\nDescribe normalised \"number_of_jobs\" column:\n{stats2}\n")


Exercise 8 - Normalise the number of jobs column

Use .describe() to find the median, upper quartile and maximum for the "number_of_jobs" column:
count    9.310000e+02
mean     3.188095e+06
std      8.058302e+06
min      4.700000e+04
25%      9.450000e+04
50%      1.570000e+05
75%      2.217000e+06
max      3.575000e+07
Name: number_of_jobs, dtype: float64


Display Median, Upper Quartile and Maximum:
Median: 157000.0
Upper Quartile: 2217000.0
Maximum: 35750000.0


Display normalised "number_of_jobs" column:
           code                  area        date  median_salary  life_satisfaction mean_salary recycling_pct  population_size  number_of_jobs  area_size  no_of_houses  borough_flag
0     E09000001        city of london  1999-12-01        33020.0                NaN       48922             0           6581.0               3        NaN           NaN             1
1     E09000002  barking and dagenham  1999-12-01        21480.0                NaN       23620             3         162

## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1

In [37]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()

# Interrogate data
# 1. Use df.describe() to find the median, upper quartile and maximum for the "number_of_jobs" column
house_stats = data["no_of_houses"].describe()

# Median
median = house_stats["50%"]

# Upper Quartile
upper_quartile = house_stats["75%"]

#  Maximum
max = house_stats["max"]

# 2. Create a new function and code to normalise the "no_of_houses" column.
def normalise_house_nos(data):
  if data["no_of_houses"] < median:
    return 1
  elif data["no_of_houses"] < upper_quartile:
    return 2
  else:
    return 3

# 3. Normalise the "no_of_houses" column by applying the function normalise_jobs. Assign it to a new column called df['housing_volume']
data["housing_volume"] = data.apply(normalise_house_nos, axis = 1)

print(f"\nDisplay normalised \"housing_volume\" column:\n{data.loc[:]}\n")



Display normalised "housing_volume" column:
           code                  area        date  median_salary  life_satisfaction mean_salary recycling_pct  population_size  number_of_jobs  area_size  no_of_houses  borough_flag  housing_volume
0     E09000001        city of london  1999-12-01        33020.0                NaN       48922             0           6581.0             NaN        NaN           NaN             1               3
1     E09000002  barking and dagenham  1999-12-01        21480.0                NaN       23620             3         162444.0             NaN        NaN           NaN             1               3
2     E09000003                barnet  1999-12-01        19568.0                NaN       23128             8         313469.0             NaN        NaN           NaN             1               3
3     E09000004                bexley  1999-12-01        18621.0                NaN       21386            18         217458.0             NaN        NaN          

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [43]:
import pandas as pd
pd.set_option('display.width', 240)

def get_csv_data():
  url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
  df = pd.read_csv(url)
  return df

data = get_csv_data()


# Interrogate data
# 1. Normalise the "area_size" column so that all values below mean are represented as 0 and otherwise are 1.
mean = data["area_size"].mean()

def area_size_normalised(data):
  if data["area_size"] < mean:
    return 0
  else:
    return 1

# 2. Assign the output to a new column called "area_size_normalised".
data["area_size_normalised"] = data.apply(area_size_normalised, axis = 1)

print(f"\nDisplay normalised \"area_size\" column:\n{data.loc[:]}\n")


count    6.660000e+02
mean     3.724903e+05
std      2.157060e+06
min      3.150000e+02
25%      2.960000e+03
50%      4.323000e+03
75%      8.220000e+03
max      1.330373e+07
Name: area_size, dtype: float64

Display normalised "area_size" column:
           code                  area        date  median_salary  life_satisfaction mean_salary recycling_pct  population_size  number_of_jobs  area_size  no_of_houses  borough_flag  area_size_normalised
0     E09000001        city of london  1999-12-01        33020.0                NaN       48922             0           6581.0             NaN        NaN           NaN             1                     1
1     E09000002  barking and dagenham  1999-12-01        21480.0                NaN       23620             3         162444.0             NaN        NaN           NaN             1                     1
2     E09000003                barnet  1999-12-01        19568.0                NaN       23128             8         313469.0             N